### ISSUE TITLE
Default format for JTFS

### ISSUE THREAD

**Issue by janden** [kymatio/kymatio#1033]
What do the NaNs in the JTFS `meta['j']` mean? Even if we have averaged (convolved) in frequency and subsampled, don't we record the central frequency here? (Same with `xi`, `sigma`, and `n` (why range?) keys.) Otherwise how can we expect to order them (e.g., for display purposes)?

**Comment by cyrusvahidi** [comment]
> What do the NaNs in the JTFS `meta['j']` mean? Even if we have averaged (convolved) in frequency and subsampled, don't we record the central frequency here? (Same with `xi`, `sigma`, and `n` (why range?) keys.) Otherwise how can we expect to order them (e.g., for display purposes)?

Could you give an example? AFAIK this is a workaround to indicate whether a path has a second order descendant (e.g if `j1 > j2`.

**Comment by lostanlen** [comment]
> Even if we have averaged (convolved) in frequency and subsampled, don't we record the central frequency here? 

The default layout is `'joint'` which means that the output is a dict/list of 3D arrays, indexed by batch ID, time, and frequency. Since there is more than one frequency involved in each array, the `n` is undefined and thus takes a NaN value. Same goes for `xi`, `sigma`and `j` which are derived from `n` by table lookup.

My proposed way to retrieve the `n`'s (plural) for a given path is to look at `n1_max` and `n1_step`. More precisely, taking a range which starts at zero and ends at `n1_max` by steps of `n1_stride` provides an iterable of `n` values, corresponding to the rows of the JTFS output in `'joint'` mode after frequency unpadding.

TL;DR this is working as intended as far as i'm aware, although it could benefit from a developer's documentation.

**Comment by janden** [comment]
> Could you give an example? AFAIK this is a workaround to indicate whether a path has a second order descendant (e.g if `j1 > j2`.

```
from kymatio.torch import TimeFrequencyScattering

S = TimeFrequencyScattering(J=8, J_fr=2, shape=8192, Q=8)

S.meta()['j']
```

> Since there is more than one frequency involved in each array, the `n` is undefined and thus takes a NaN value.

I'm not sure I agree with this. By the same token, the time axis would be undefined in the standard scattering since multiple time samples are involved when taking a convolution.

We do still have a notion of frequency since we perform a *convolution* along that axis. We then subsample along that axis, but that does not preclude storing those frequencies in the meta information. IIRC this is what was done in [ScatNet](https://github.com/scatnet/scatnet/blob/59d935afa20359845282a3518134e24244862c1f/core/joint_tf_wavelet_layer_1d.m#L240).

> My proposed way to retrieve the `n`'s (plural) for a given path is to look at `n1_max` and `n1_step`. More precisely, taking a range which starts at zero and ends at `n1_max` by steps of `n1_stride` provides an iterable of `n` values, corresponding to the rows of the JTFS output in `'joint'` mode after frequency unpadding. From this iterable of `n`'s, it is easy to derive `xi`'s et al.

I'm not following here. Why do we have a stride along the frequency axis here? Wouldn't one JTFS coefficient correspond to a *contiguous* range of frequencies, if anything? I also don't see how this would enable a simple way of ordering the coefficients by `j1` for display purposes.

Maybe I'm missing something fundamental here…

**Comment by lostanlen** [comment]
> We do still have a notion of frequency since we perform a convolution along that axis. We then subsample along that axis, but that does not preclude storing those frequencies in the meta information. IIRC this is what was done in [ScatNet](https://github.com/scatnet/scatnet/blob/59d935afa20359845282a3518134e24244862c1f/core/joint_tf_wavelet_layer_1d.m#L240).

Set `format='time'` in Kymatio JTFS to replicate the layout of ScatNet. The `n` and friends will (should be) be there.

**Comment by janden** [comment]
Ok. Is there an explanation of these various formats somewhere that I can look at?

**Comment by janden** [comment]
Never mind found it: https://github.com/kymatio/kymatio/blob/ca43df55be70deb11e0ab166c1195b8e614ab04d/kymatio/scattering1d/frontend/base_frontend.py#L681

Question is then what the default format should be. I'm guessing you favor `joint`?

**Comment by lostanlen** [comment]
Yes, i tend to favor joint formatting. I expect joint formatting to allocate less memory (no need to copy 3D arrays into 2D arrays) and thus a bit faster. It's also more in line with BatchNorm2D and Conv2D, which are natural candidates for JTFS postprocessing. The only use case i see for time formatting is if you want to group paths by same xi1.

**Comment by janden** [comment]
Well the counterargument here would be to conform with the output format of `Scattering1D`. The fact that so many of our metadata becomes irrelevant suggests that `format='array'` would be a more natural default.

> I expect joint formatting to allocate less memory (no need to copy 3D arrays into 2D arrays) and thus a bit faster.

Would be good to see some numbers on this. I'm not sure.

> It's also more in line with BatchNorm2D and Conv2D, which are natural candidates for JTFS postprocessing. 

You could make that argument for `Scattering1d` as well. Is the output a series of feature vectors in time or is it a time–frequency image.

> The only use case i see for time formatting is if you want to group paths by same xi1.

Well again, anytime we view our output as a time series of some sort, it makes sense. Could be classification, segmentation, activity detection, etc. 1D in – 1D out.

@cyrusvahidi @eickenberg @MuawizChaudhary What are you thoughts here?

**Comment by janden** [comment]
> What do you mean by `format='array'`?

Sorry meant `format='time'` here.

> as far as i can see, `format='time'` is derived from `format='joint'` so it can't be any faster.

Not saying it would be faster, just that I'm not sure the difference in timing is big enough to justify it.

> I think we're in agreement on the implementation, and that the current discussion is on what should be the default, am i right? If so, that should resolve the issue as to why there are some NaN's in JTFS meta under `format='joint'`.

Yes and yes.

> The issue has moved to whether that should be the default behavior of not. I can have it either way

Right. I will do some timing tests to compare, then we can make a decision.

**Comment by janden** [comment]
Have the following result on a 20-core Xeon 2.4 GHz
```
In [12]: S = TimeFrequencyScattering(J=10, J_fr=2, shape=N, Q=8, format="joint")                                                                                         
                                                                                                                                                                         
In [13]: %timeit S(x)                                                                                                                                                    
753 ms ± 13.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)                                                                                                      
                                                                                                                                                                         
In [14]: S = TimeFrequencyScattering(J=10, J_fr=2, shape=N, Q=8, format="time")                                                                                          
                                                                           
In [15]: %timeit S(x)                                                                                                                                                    
765 ms ± 9.95 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)    
```

With CUDA on a V100, I get
```
In [17]: S = TimeFrequencyScattering(J=10, J_fr=2, shape=N, Q=8, format="joint").cuda()                                                                                  
                                                                                                                                                                         
In [18]: %timeit S(x)                           
75.4 ms ± 653 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)
                                                           
In [19]: S = TimeFrequencyScattering(J=10, J_fr=2, shape=N, Q=8, format="time").cuda()   
                                                                                         
In [20]: %timeit S(x)                                                                    
86.1 ms ± 290 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)  
```
So `format="time"` is slightly slower here. Note that this is for `J_fr = 2`, for larger values of `J_fr` (say `J_fr = 5`), there is virtually no difference (since there are fewer remaining frequencies to realign).

I don't think the difference here is big enough to warrant a deviation from the behavior of `Scattering1D` (that is, `format="time"`). Then we can mention in the docs that `format="joint"` is slightly faster since it involves less postprocessing if we want.

**Comment by lostanlen** [comment]
OK! i'm sold on `format='time'` by default
Thanks for these benchmarks

**Comment by janden** [comment]
Great! I'll put together a PR for this and #1032.

**Comment by MuawizChaudhary** [comment]
Closing as this was resolved by PR #1041 which has since been merged. Please reopen if I am wrong.
